### PR TITLE
Update CmsPageRouter.php

### LINKED ISSUE #1655: CmsPageRouter not working
#### Sonata packages

<details><summary>show</summary>
<p>

```
$ composer show --latest 'sonata-project/*'
Info from https://repo.packagist.org: #StandWithUkraine
Color legend:
- patch or minor release available - update recommended
- major release available - update possible
- up to date version

Direct dependencies required in composer.json:
sonata-project/admin-bundle              4.20.0 4.22.2 The missing Symfony Admin Generator
sonata-project/doctrine-orm-admin-bundle 4.8.0  4.9.1  Integrate Doctrine ORM into the SonataAdminBundle
sonata-project/media-bundle              4.5.1  4.5.1  Symfony SonataMediaBundle
sonata-project/page-bundle               4.0.0  4.0.0  This bundle provides a Site and Page management through container and block services

Transitive dependencies not required in composer.json:
sonata-project/block-bundle              4.19.0 4.19.0 Symfony SonataBlockBundle
sonata-project/cache                     2.2.0  2.2.0  Cache library
Package sonata-project/cache is abandoned, you should avoid using it. No replacement was suggested.
sonata-project/doctrine-extensions       1.18.1 2.1.0  Doctrine2 behavioral extensions
sonata-project/exporter                  3.0.0  3.1.1  Lightweight Exporter library
sonata-project/form-extensions           1.18.0 1.18.0 Symfony form extensions
sonata-project/seo-bundle                3.5.0  3.5.0  Symfony SonataSeoBundle
sonata-project/twig-extensions           1.12.0 2.0.0  Sonata twig extensions
```

</p>
</details>

#### Symfony packages

<details><summary>show</summary>
<p>

```
$ composer show --latest 'symfony/*'
Color legend:
- patch or minor release available - update recommended
- major release available - update possible
- up to date version

Direct dependencies required in composer.json:
symfony/console                    v5.4.15 v6.2.3  Eases the creation of beautiful and testable command line interfaces
symfony/debug-bundle               v5.4.11 v6.2.1  Provides a tight integration of the Symfony VarDumper component and the ServerLogCommand from MonologBridge into the Symfon...
symfony/dependency-injection       v5.4.13 v6.2.3  Allows you to standardize and centralize the way objects are constructed in your application
symfony/dotenv                     v5.4.5  v6.2.0  Registers environment variables from a .env file
symfony/flex                       v1.19.3 v2.2.4  Composer plugin for Symfony
symfony/framework-bundle           v5.4.14 v6.2.3  Provides a tight integration between Symfony components and the Symfony full-stack framework
symfony/mailer                     v5.4.15 v6.2.2  Helps sending emails
symfony/maker-bundle               v1.47.0 v1.48.0 Symfony Maker helps you create empty commands, controllers, form classes, tests and more so you can forget about writing bo...
symfony/monolog-bundle             v3.8.0  v3.8.0  Symfony MonologBundle
symfony/proxy-manager-bridge       v5.4.6  v6.2.0  Provides integration for ProxyManager with various Symfony components
symfony/runtime                    v5.4.11 v6.2.0  Enables decoupling PHP applications from global state
symfony/security-bundle            v5.4.11 v6.2.3  Provides a tight integration of the Security component into the Symfony full-stack framework
symfony/serializer                 v5.4.15 v6.2.3  Handles serializing and deserializing data structures, including object graphs, into array structures or other formats like...
symfony/stopwatch                  v5.4.13 v6.2.0  Provides a way to profile code
symfony/twig-bundle                v5.4.8  v6.2.3  Provides a tight integration of Twig into the Symfony full-stack framework
symfony/ux-autocomplete            v2.4.0  v2.6.1  JavaScript Autocomplete functionality for Symfony
symfony/ux-dropzone                v2.4.0  v2.6.1  File input dropzones for Symfony Forms
symfony/ux-live-component          v2.4.0  v2.6.1  Live components for Symfony
symfony/ux-swup                    v2.4.0  v2.6.1  Swup integration for Symfony
symfony/ux-twig-component          v2.4.0  v2.6.1  Twig components for Symfony
symfony/validator                  v5.4.15 v6.2.3  Provides tools to validate values
symfony/web-profiler-bundle        v5.4.14 v6.2.4  Provides a development tool that gives detailed information about the execution of any request
symfony/webpack-encore-bundle      v1.16.0 v1.16.0 Integration with your Symfony app & Webpack Encore!
symfony/yaml                       v5.4.14 v6.2.2  Loads and dumps YAML files

Transitive dependencies not required in composer.json:
symfony/asset                      v5.4.13 v6.2.0  Manages URL generation and versioning of web assets such as CSS stylesheets, JavaScript files and image files
symfony/cache                      v5.4.15 v6.2.4  Provides extended PSR-6, PSR-16 (and tags) implementations
symfony/cache-contracts            v2.5.2  v3.2.0  Generic abstractions related to caching
symfony/config                     v5.4.11 v6.2.0  Helps you find, load, combine, autofill and validate configuration values of any kind
symfony/css-selector               v5.4.11 v6.2.3  Converts CSS selectors to XPath expressions
symfony/deprecation-contracts      v3.1.1  v3.2.0  A generic function and convention to trigger deprecation notices
symfony/doctrine-bridge            v5.4.15 v6.2.3  Provides integration for Doctrine with various Symfony components
symfony/error-handler              v5.4.15 v6.2.3  Provides tools to manage errors and ease debugging PHP code
symfony/event-dispatcher           v5.4.9  v6.2.2  Provides tools that allow your application components to communicate with each other by dispatching events and listening to...
symfony/event-dispatcher-contracts v3.1.1  v3.2.0  Generic abstractions related to dispatching event
symfony/expression-language        v5.4.14 v6.2.2  Provides an engine that can compile and evaluate expressions
symfony/filesystem                 v5.4.13 v6.2.0  Provides basic utilities for the filesystem
symfony/finder                     v5.4.11 v6.2.3  Finds files and directories via an intuitive fluent interface
symfony/form                       v5.4.15 v6.2.3  Allows to easily create, process and reuse HTML forms
symfony/http-client                v5.4.15 v6.2.2  Provides powerful methods to fetch HTTP resources synchronously or asynchronously
symfony/http-client-contracts      v2.5.2  v3.2.0  Generic abstractions related to HTTP clients
symfony/http-foundation            v5.4.15 v6.2.2  Defines an object-oriented layer for the HTTP specification
symfony/http-kernel                v5.4.15 v6.2.4  Provides a structured process for converting a Request into a Response
symfony/intl                       v5.4.15 v6.2.0  Provides a PHP replacement layer for the C intl extension that includes additional data from the ICU library
symfony/mime                       v5.4.14 v6.2.2  Allows manipulating MIME messages
symfony/monolog-bridge             v5.4.10 v6.2.2  Provides integration for Monolog with various Symfony components
symfony/options-resolver           v5.4.11 v6.2.0  Provides an improved replacement for the array_replace PHP function
symfony/password-hasher            v5.4.11 v6.2.0  Provides password hashing utilities
symfony/polyfill-intl-grapheme     v1.26.0 v1.27.0 Symfony polyfill for intl's grapheme_* functions
symfony/polyfill-intl-icu          v1.26.0 v1.27.0 Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-intl-idn          v1.26.0 v1.27.0 Symfony polyfill for intl's idn_to_ascii and idn_to_utf8 functions
symfony/polyfill-intl-normalizer   v1.26.0 v1.27.0 Symfony polyfill for intl's Normalizer class and related functions
symfony/polyfill-mbstring          v1.26.0 v1.27.0 Symfony polyfill for the Mbstring extension
symfony/polyfill-php73             v1.26.0 v1.27.0 Symfony polyfill backporting some PHP 7.3+ features to lower PHP versions
symfony/polyfill-php80             v1.26.0 v1.27.0 Symfony polyfill backporting some PHP 8.0+ features to lower PHP versions
symfony/polyfill-php81             v1.26.0 v1.27.0 Symfony polyfill backporting some PHP 8.1+ features to lower PHP versions
symfony/process                    v5.4.11 v6.2.0  Executes commands in sub-processes
symfony/property-access            v5.4.15 v6.2.3  Provides functions to read and write from/to an object or array using a simple string notation
symfony/property-info              v5.4.15 v6.2.3  Extracts information about PHP class' properties using metadata of popular sources
symfony/routing                    v5.4.15 v6.2.3  Maps an HTTP request to a set of configuration variables
symfony/security-acl               v3.3.2  v3.3.2  Symfony Security Component - ACL (Access Control List)
symfony/security-core              v5.4.15 v6.2.2  Symfony Security Component - Core Library
symfony/security-csrf              v5.4.11 v6.2.0  Symfony Security Component - CSRF Library
symfony/security-guard             v5.4.13 v5.4.13 Symfony Security Component - Guard
symfony/security-http              v5.4.15 v6.2.2  Symfony Security Component - HTTP Integration
symfony/service-contracts          v2.5.2  v3.2.0  Generic abstractions related to writing services
symfony/string                     v5.4.15 v6.2.2  Provides an object-oriented API to strings and deals with bytes, UTF-8 code points and grapheme clusters in a unified way
symfony/templating                 v5.4.11 v6.2.0  Provides all the tools needed to build any kind of template system
symfony/translation                v5.4.14 v6.2.3  Provides tools to internationalize your application
symfony/translation-contracts      v2.5.2  v3.2.0  Generic abstractions related to translation
symfony/twig-bridge                v5.4.14 v6.2.3  Provides integration for Twig with various Symfony components
symfony/var-dumper                 v5.4.14 v6.2.3  Provides mechanisms for walking through any arbitrary PHP variable
symfony/var-exporter               v5.4.10 v6.2.3  Allows exporting any serializable PHP data structure to plain PHP code
```

</p>
</details>

#### PHP version

```
$ php -v
PHP 8.1.10 (cli) (built: Sep  1 2022 21:43:31) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.10, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.10, Copyright (c), by Zend Technologies
    with Xdebug v3.1.5, Copyright (c) 2002-2022, by Derick Rethans
```

## Subject

The CmsPageRouter seems to not work any more when using a page as first parameter of generate.

## Steps to reproduce

Create a page entity and use 

```php
<?php

/** @var PageInterface */
$page = '...';
$this->generateUrl($page);
```

or in twig template
```twig
{{ path(page) }}
```

After searching a litte into symfony documentation, i found this:
```php
<?php

namespace Symfony\Cmf\Component\Routing;

//...
class ChainRouter
{
//...
    /**
     * {@inheritdoc}
     *
     * @param mixed $name
     *
     * The CMF routing system used to allow to pass route objects as $name to generate the route.
     * Since Symfony 5.0, the UrlGeneratorInterface declares $name as string. We widen the contract
     * for BC but deprecate passing non-strings.
     * Instead, Pass the RouteObjectInterface::OBJECT_BASED_ROUTE_NAME as route name and the object
     * in the parameters with key RouteObjectInterface::ROUTE_OBJECT.
     *
     * Loops through all registered routers and returns a router if one is found.
     * It will always return the first route generated.
     */
    public function generate($name, $parameters = [], $absolute = UrlGeneratorInterface::ABSOLUTE_PATH)
    {
    }
}
```

So may be the goog way to genreate page url is by doing

```php
$this->generateUrl(RouteObjectInterface::OBJECT_BASED_ROUTE_NAME, [
     RouteObjectInterface::ROUTE_OBJECT => $page
]);

```
and in twig:

```twig
{{ path('cmf_routing_object', {
     '_route_object': page
}) }}
```

## Expected results

The page page path url

## Actual results

```
TypeError:
Symfony\Bundle\FrameworkBundle\Controller\AbstractController::generateUrl(): Argument #1 ($route) must be of type string, App\Entity\SonataPage given, called in /srv/app/src/Controller/Admin/TestController.php on line 69

  at vendor/symfony/framework-bundle/Controller/AbstractController.php:140
  at Symfony\Bundle\FrameworkBundle\Controller\AbstractController->generateUrl(object(SonataPage))
     (src/Controller/Admin/TestController.php:69)
  at App\Controller\Admin\TestController->index()
     (vendor/symfony/http-kernel/HttpKernel.php:163)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:75)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (vendor/symfony/runtime/Runner/Symfony/HttpKernelRunner.php:35)
  at Symfony\Component\Runtime\Runner\Symfony\HttpKernelRunner->run()
     (vendor/autoload_runtime.php:35)
  at require_once('/srv/app/vendor/autoload_runtime.php')
     (public/index.php:5)   
```

### PR DESCRIPTION
**Pull request by tdumalin**
Add OBJECT_BASED_ROUTE_NAME feature

## Subject

Fix for the CmsPageRouter when generating url from page entity

I am targeting this branch, because this is the last stable version

Closes #1655.

## Changelog

```markdown

### Deprecated

- Add a TODO comment to remove a statement which is deprecated by the symfony url generator 

### Removed

### Fixed
- A bug that prevent to render page url using the entity has first parameter of urlGenerator generate method by updating the `Sonata\PageBundle\Route\CmsPageRouter::generate()` method

## To do

- [ ] Update the tests;
- [ ] Update the documentation;

### CODE REVIEW

**Comment by VincentLanglet** [comment]
Hi @tdumalin, can you detail the issue ?

There is the code:
```
if ($name instanceof PageInterface) {
                $url = $this->generateFromPage($name, $parameters, $referenceType);
            }
```
with 
```
private function generateFromPage(PageInterface $page, array $parameters = [], int $referenceType = self::ABSOLUTE_PATH): string
    {
        // hybrid pages use, by definition, the default routing mechanism
        if ($page->isHybrid()) {
            $routeName = $page->getRouteName();

            if (null === $routeName) {
                throw new \RuntimeException('Page is hybrid but has no route name');
            }

            return $this->router->generate($routeName, $parameters, $referenceType);
        }

        $url = $page->getUrl();

        if (null === $url) {
            throw new \RuntimeException(sprintf('Page "%d" has no url.', $page->getId() ?? ''));
        }

        if (!$this->context instanceof SiteRequestContextInterface) {
            throw new \RuntimeException('The context must be an instance of SiteRequestContextInterface');
        }

        // Get current site
        $currentSite = $this->context->getSite();
        // Change to new site
        $this->context->setSite($page->getSite());
        // Fetch Url
        $decoratedUrl = $this->decorateUrl($url, $parameters, $referenceType);
        // Swap back to original site
        $this->context->setSite($currentSite);

        return $decoratedUrl;
    }
```

So we're supposed to support PageInterface, and I don't see why we should remove the support.

None of the stacktrace in https://github.com/sonata-project/SonataPageBundle/issues/1655#issue-1529148113 is talking about sonata page

**Comment by tdumalin** [comment]
Hi @VincentLanglet ,

Check the #1655 issue, the current behavior throw an error since the parameter $name of UrlGeneratorInterface::generate method is typed as a string.

https://github.com/symfony/symfony/blob/6.3/src/Symfony/Component/Routing/Generator/UrlGeneratorInterface.php

so we should not set this parameter with an object (entity) but with a string and a custom parameter as explained in the issue.

Let me know if you need more info,

**Comment by VincentLanglet** [comment]
> Hi @VincentLanglet ,
> 
> Check the #1655 issue, the current behavior throw an error since the parameter $name of UrlGeneratorInterface::generate method is typed as a string.

It doesn't forbid to support a wilder range of types:
https://onlinephp.io/c/bcc49

> so we should not set this parameter with an object (entity) but with a string and a custom parameter as explained in the issue.

Like I said, the error stack trace is coming from symfony and not from our class. I would say you use the wrong class/method.

**Comment by tdumalin** [comment]
My bad i actually paste the stack trace when generating an url from a controller so type hint is triggered, here's the stacktrace when using path twig method.

```
Twig\Error\RuntimeError:
An exception has been thrown during the rendering of a template ("None of the chained routers were able to generate route: Route 'Réalisations' not found").

  at templates/admin/test/index.html.twig:1
  at Twig\Template->displayWithErrorHandling(array('page' => object(SonataPageSnapshot), 'wrap_fields_with_addons' => true, 'app' => object(AppVariable),  'sonata_block' => object(GlobalVariables), 'sonata_config' => object(SonataConfiguration), 'sonata_page' => object(GlobalVariables), 'sonata_page_admin' => object(PageAdmin), 'sonata_media' => object(GlobalVariables)), array())
     (vendor/twig/twig/src/Template.php:378)
  at Twig\Template->display(array('page' => object(SonataPageSnapshot)))
     (vendor/twig/twig/src/Template.php:390)
  at Twig\Template->render(array('page' => object(SonataPageSnapshot)), array())
     (vendor/twig/twig/src/TemplateWrapper.php:45)
  at Twig\TemplateWrapper->render(array('page' => object(SonataPageSnapshot)))
     (vendor/twig/twig/src/Environment.php:318)
  at Twig\Environment->render('admin/test/index.html.twig', array('page' => object(SonataPageSnapshot)))
     (vendor/symfony/framework-bundle/Controller/AbstractController.php:258)
  at Symfony\Bundle\FrameworkBundle\Controller\AbstractController->renderView('admin/test/index.html.twig', array('page' => object(SonataPageSnapshot)))
     (vendor/symfony/framework-bundle/Controller/AbstractController.php:266)
  at Symfony\Bundle\FrameworkBundle\Controller\AbstractController->render('admin/test/index.html.twig', array('page' => object(SonataPageSnapshot)))
     (src/Controller/Admin/TestController.php:73)
  at App\Controller\Admin\TestController->index(object(SnapshotManager))
     (vendor/symfony/http-kernel/HttpKernel.php:163)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:75)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (vendor/symfony/runtime/Runner/Symfony/HttpKernelRunner.php:35)
  at Symfony\Component\Runtime\Runner\Symfony\HttpKernelRunner->run()
     (vendor/autoload_runtime.php:35)
  at require_once('/srv/app/vendor/autoload_runtime.php')
     (public/index.php:5)  
```

Actually what really happened is that the page entity is casted into a string at some point and the router try to generate an url where the route is the page name (because the `Sonata\PageBundle\Model\Page::__toString` method return page name)

**Comment by VincentLanglet** [comment]
This is because of twig methods.
```
public function getPath(string $name, array $parameters = [], bool $relative = false): string
    {
        return $this->generator->generate($name, $parameters, $relative ? UrlGeneratorInterface::RELATIVE_PATH : UrlGeneratorInterface::ABSOLUTE_PATH);
    }

    public function getUrl(string $name, array $parameters = [], bool $schemeRelative = false): string
    {
        return $this->generator->generate($name, $parameters, $schemeRelative ? UrlGeneratorInterface::NETWORK_PATH : UrlGeneratorInterface::ABSOLUTE_URL);
    }
```

This is not because it doesn't work with twig method that we have to remove the Php feature.

We should introduce two twig extension:
```
page_path()
page_url()
```
instead and update the doc.

**Comment by tdumalin** [comment]
Hi there,

I've pushed some changes according to your comments, by creating two twig functions to render page url from templates.
I also removed the TODO comment in the `CmsPageRouter`  comment to keep the current implementations.

But i think you should add the `cmf_routing_object` implementations as it doesn't change any thing from the current behavior and it's also the recommended way to generate url from object according to symfony documentation.

With this implementation it's also possible to use the default controller method
`Symfony\Bundle\FrameworkBundle\Controller\AbstractController::generateUrl()`

Or use an autowired typed service:
`Symfony\Component\Routing\RouterInterface`
`Symfony\Component\Routing\Generator\UrlGeneratorInterface`

to generate page url 

Keep me in touch,

**Comment by haivala** [comment]
I'm wondering: Do these new twig functions help with https://github.com/sonata-project/SonataPageBundle/issues/1616 ?

File: src/Twig/Extension/PageExtension.php [hunk 1]
```diff
@@ -218,4 +221,30 @@ public function controller(string $controller, array $attributes = [], array $qu
...
+     * @param array $parameters
+     * @return string
+     */
+    public function pagePath(PageInterface $page, array $parameters = []) : string
+    {
+        $parameters = array_merge($parameters, [
+            RouteObjectInterface::ROUTE_OBJECT => $page
+        ]);
+        return $this->router->generate(RouteObjectInterface::OBJECT_BASED_ROUTE_NAME, $parameters, RouterInterface::ABSOLUTE_PATH);
+    }
```
Comment (VincentLanglet) [inline src/Twig/Extension/PageExtension.php:230]: The implementation I had in mind was:
```
    /**
     * @param string|PageInterface $name
     * @param array<mixed>         $parameters
     */
    public function getPagePath(string $name, array $parameters = [], bool $relative = false): string
    {
        return $this->cmsPageRouter->generate($name, $parameters, $relative ? UrlGeneratorInterface::RELATIVE_PATH : UrlGeneratorInterface::ABSOLUTE_PATH);
    }

    /**
     * @param string|PageInterface $name
     * @param array<mixed>         $parameters
     */
    public function getPageUrl(string $name, array $parameters = [], bool $schemeRelative = false): string
    {
        return $this->cmsPageRouter->generate($name, $parameters, $schemeRelative ? UrlGeneratorInterface::NETWORK_PATH : UrlGeneratorInterface::ABSOLUTE_URL);
    }
```

This way, you get the same behavior than before, and no need for OBJECT_BASED_ROUTE_NAME

**Review by VincentLanglet — CHANGES_REQUESTED** [review]

File: src/Route/CmsPageRouter.php [hunk 2]
```diff
@@ -19,7 +19,9 @@
 use Sonata\PageBundle\Request\SiteRequestContextInterface;
 use Sonata\PageBundle\Site\SiteSelectorInterface;
 use Symfony\Cmf\Component\Routing\ChainedRouterInterface;
+use Symfony\Cmf\Component\Routing\RouteObjectInterface;
```
Comment (eerison) [inline src/Route/CmsPageRouter.php:22]: is it possible, we don't use class/interfaces from `Symfony\Cmf\Component\Routing` ?
we should keep using one routing component and IMO we should keep using things from `symfony routing` only!

File: src/Twig/Extension/PageExtension.php [hunk 3]
```diff
@@ -218,4 +221,30 @@ public function controller(string $controller, array $attributes = [], array $qu
 
         return HttpKernelExtension::controller($controller, $attributes, $query);
     }
+    
+    /**
+     * @param PageInterface $page
+     * @param array $parameters
+     * @return string
+     */
+    public function pageUrl(PageInterface $page, array $parameters = []) : string
```
Comment (eerison) [inline src/Twig/Extension/PageExtension.php:211]: IMO you should create test for those 2 new methods!

File: src/Route/CmsPageRouter.php (hunk 2 above)
Comment (tdumalin) [reply src/Route/CmsPageRouter.php:22]: I don't understand your point, `RouteObjectInterface` is only defined in `Symfony\Cmf`, you prefer to use string (cmf_routing_object)  value instead on class constant ?

File: src/Twig/Extension/PageExtension.php (hunk 3 above)
Comment (tdumalin) [reply src/Twig/Extension/PageExtension.php:211]: So what should i do ? Use the`cmsPageRouter as @VincentLanglet  proposed or keep the cmf_routing_object and create tests ?

File: src/Route/CmsPageRouter.php (hunk 2 above)
Comment (VincentLanglet) [reply src/Route/CmsPageRouter.php:22]: I don't understand either.

Comment (eerison) [reply src/Route/CmsPageRouter.php:22]: No, I mean, we don't depends of Symfony cmf anymore!

File: src/Twig/Extension/PageExtension.php (hunk 3 above)
Comment (eerison) [reply src/Twig/Extension/PageExtension.php:211]: I'm not saying to use hardcoded string `cmf_routing_object`!
and any solution that you choose you should write test!

Comment (VincentLanglet) [reply src/Twig/Extension/PageExtension.php:211]: I'll recommend to follow my implementation and confirm it works for you.

Best would be to have a functional test of the twig method
````
{{ page_path($page) }}
```

File: src/Route/CmsPageRouter.php (hunk 2 above)
Comment (eerison) [reply src/Route/CmsPageRouter.php:22]: I mean

> is it possible, we don't use class/interfaces from Symfony\Cmf\Component\Routing ?

sorry 😄

File: src/Twig/Extension/PageExtension.php (hunk 3 above)
Comment (tdumalin) [reply src/Twig/Extension/PageExtension.php:211]: Ok, i'll do my best to answer your demand, when i have some free time I go back on the subject

**Pull request closed without merging**